### PR TITLE
Support VK_KHR_dynamic_rendering_local_read

### LINKED ISSUE #3341: Please consider adding support for VK_KHR_dynamic_rendering_local_read
## Description

The VK_KHR_dynamic_rendering_local_read extension closes a big gap between VK_KHR_dynamic_rendering and VkRenderPass objects. For applications that use input attachments, this extension is an absolute necessity to enable them to move to dynamic rendering (which is already part of core Vulkan 1.3).

Please consider adding support for this extension. It's likely not too difficult to implement, but it does include an annoyance with color and input attachment index mapping. In particular, unlike the subpass mapping of color locations to render pass attachments, the mapping in this extension does not affect blend states etc matching dynamic rendering attachments.

## Environment

* RenderDoc version: v1.33
* Operating System: Linux
* Graphics API: Vulkan

### PR DESCRIPTION
**Pull request by ShabbyX**
## Description

VK_KHR_dynamic_rendering_local_read (DRLR) closes the gap left by VK_KHR_dynamic_rendering (DR) where input attachments couldn't be used. As the Vulkan API moves towards dynamic rendering, usage of DRLR will increase and DR will eventually replace now-legacy VkRenderPass objects.

This change adds support for DRLR to RenderDoc. At a high level, DRLR allows the render pass color attachments to be mapped to "locations". Since the rest of the API does not use these remapped locations (notably, blend and vkCmdClearAttachments use the attachment index), I opted for RenderDoc to continue referencing the attachment by its index in the UI. However, where location matters (i.e. connection to the shader), the mapping is specified in the UI:

image

Additionally, an attachment can be mapped to VK_ATTACHMENT_UNUSED, in which case writes to it are automatically disabled. In this case, RenderDoc would tag the attachment as `[disabled]` and automatically set the Write Mask to `____`:

image

DRLR additionally includes a mapping of attachments to `InputAttachmentIndex` decorations in the shader. For depth/stencil, the shader is no longer required to specify this decoration at all, in which case the corresponding API (`VkRenderingInputAttachmentIndexInfoKHR`) would also specify `nullptr` for the mapping. If VK_ATTACHMENT_UNUSED is specified in the mapping, the attachment is not expected to be used as input attachment. In the code, an additional flag for depth and stencil is present to differentiate between `nullptr` and `VK_ATTACHMENT_UNUSED`, with the format being an "implicit" mapping.

Notably, RenderDoc does not actually care about input attachment indices, but rather uses set/binding information to read from input attachments (e.g. during shader debugging), so the input index mapping information only ends up being used for the html export from the pipeline tab.

Fixes #3341

### CODE REVIEW

**Comment by ShabbyX** [comment]
@Themaister the Pipeline State tab also has an "Export to Fossilize database" functionality. Is there something I need to manually implement for that?

File: renderdoc/driver/vulkan/extension_support.md [hunk 1]
```diff
@@ -241,9 +242,9 @@ KHR extensions will definitely be implemented at some point, though KHR extensio
 ## KHR Extensions
 
 * `VK_KHR_cooperative_matrix`
-* `VK_KHR_dynamic_rendering_local_read`
 * `VK_KHR_maintenance5`
 * `VK_KHR_maintenance6`
+* `VK_KHR_maintenance7`
```
Comment (baldurk) [inline renderdoc/driver/vulkan/extension_support.md]: Please do not add extensions out of band to this list, this will be updated when the next header update happens.

File: renderdoc/api/replay/vk_pipestate.h [hunk 2]
```diff
@@ -859,6 +859,30 @@ samples used to render this subpass.
...
+  rdcarray<uint32_t> colorAttachmentInputIndices;
+
+  DOCUMENT("Whether or not depth input attachment index is implicit (dynamic rendering).");
+  bool isDepthInputAttachmentIndexImplicit = true;
+
+  DOCUMENT("Whether or not stencil  input attachment index is implicit (dynamic rendering).");
+  bool isStencilInputAttachmentIndexImplicit = true;
+
+  DOCUMENT("Depth input attachment index if explicit (dynamic rendering).");
+  uint32_t depthInputAttachmentIndex = 0x1234ABCD;
```
Comment (baldurk) [inline renderdoc/api/replay/vk_pipestate.h]: This magic number doesn't seem to make any sense to me. From a quick look it seems like this value is only valid if `isDepthInputAttachmentIndexImplicit` is false? in which case this should default to 0.

File: renderdoc/driver/vulkan/vk_common.h [hunk 3]
```diff
@@ -1053,6 +1053,8 @@ enum class VulkanChunk : uint32_t
   vkCmdSetVertexInputEXT,
   vkCmdBeginRendering,
   vkCmdEndRendering,
+  vkCmdSetRenderingAttachmentLocationsKHR,
+  vkCmdSetRenderingInputAttachmentIndicesKHR,
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_common.h]: You can't add new elements in the middle of an enum. This will renumber the enum and cause a backwards-incompatible break of serialisation for all functions after that.

File: renderdoc/driver/vulkan/vk_pixelhistory.cpp [hunk 4]
```diff
@@ -3707,7 +3722,6 @@ struct VulkanPixelHistoryPerFragmentCallback : VulkanPixelHistoryCallback
     else
     {
       pipes.primitiveIdPipe = VK_NULL_HANDLE;
-      RDCWARN("Can't get primitive ID at event %u due to lack of geometry shader support", eid);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_pixelhistory.cpp:3710]: Please do not remove this warning.

File: renderdoc/driver/vulkan/vk_pixelhistory.cpp [hunk 5]
```diff
@@ -3317,6 +3319,13 @@ struct VulkanPixelHistoryPerFragmentCallback : VulkanPixelHistoryCallback
           }
         }
 
+        // The primitive ID shader always writes to location 0, so make sure that's not remapped
+        // with VK_KHR_dynamic_rendering_local_read.
+        if(isPrimitiveIDPipe && prevState.dynamicRendering.localRead.AreLocationsNonDefault())
+        {
+          state.dynamicRendering.localRead = DynamicRenderingLocalRead{};
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_pixelhistory.cpp]: Please use a normal constructor syntax `DynamicRenderingLocalRead()` or just `{}` if you prefer.

File: renderdoc/driver/vulkan/vk_pixelhistory.cpp [hunk 6]
```diff
@@ -3329,7 +3338,13 @@ struct VulkanPixelHistoryPerFragmentCallback : VulkanPixelHistoryCallback
         m_pDriver->ReplayDraw(cmd, *action);
         state.EndRenderPass(cmd);
 
-        if(i == 1)
+        // Restore location mapping for the other pipelines.
+        if(isPrimitiveIDPipe && prevState.dynamicRendering.localRead.AreLocationsNonDefault())
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_pixelhistory.cpp]: You could simplify this to unconditionally restore the state - it's not that expensive and if it hasn't been changed that's a no-op, and it saves any issues if the conditions for save and restore get changed and out of sync in future.

File: renderdoc/driver/vulkan/vk_next_chains.cpp [hunk 7]
```diff
@@ -1593,6 +1592,30 @@ size_t GetNextPatchSize(const void *pNext)
         memSize += info->attachmentCount * sizeof(VkImageView);
         break;
       }
+      case VK_STRUCTURE_TYPE_RENDERING_ATTACHMENT_LOCATION_INFO_KHR:
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_next_chains.cpp]: Because these arrays are simple integers and don't contain handles, you don't have to do the complex path of allocating these. This is only for when we need to duplicate memory in order to unwrap handles because we can't unwrap in-place in the application's memory. Effectively we only copy deep enough to the deepest handle, everything else we can leave as-is pointing into application memory for these arrays.

File: renderdoc/driver/vulkan/vk_replay.cpp [hunk 8]
```diff
@@ -1942,6 +1942,17 @@ void VulkanReplay::SavePipelineState(uint32_t eventId)
...
+    ret.currentPass.renderpass.colorAttachmentLocations = dyn.localRead.colorAttachmentLocations;
+    ret.currentPass.renderpass.colorAttachmentInputIndices =
+        dyn.localRead.colorAttachmentInputIndices;
+    ret.currentPass.renderpass.isDepthInputAttachmentIndexImplicit =
+        dyn.localRead.isDepthInputAttachmentIndexImplicit;
+    ret.currentPass.renderpass.isStencilInputAttachmentIndexImplicit =
+        dyn.localRead.isStencilInputAttachmentIndexImplicit;
+    ret.currentPass.renderpass.depthInputAttachmentIndex = dyn.localRead.depthInputAttachmentIndex;
+    ret.currentPass.renderpass.stencilInputAttachmentIndex =
+        dyn.localRead.stencilInputAttachmentIndex;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_replay.cpp:1955]: These should be reset to default values in the `else` because the pipeline state is updated in place, it's not recreated every event. Without being reset this will linger around and still be set when a different event is selected that doesn't use dynamic rendering.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 9]
```diff
@@ -1344,6 +1355,19 @@ bool WrappedVulkan::Serialise_vkBeginCommandBuffer(SerialiserType &ser, VkComman
...
+      {
+        const VkRenderingAttachmentLocationInfoKHR *dynAttachmentLocations =
+            (const VkRenderingAttachmentLocationInfoKHR *)FindNextStruct(
+                BeginInfo.pInheritanceInfo, VK_STRUCTURE_TYPE_RENDERING_ATTACHMENT_LOCATION_INFO_KHR);
+        const VkRenderingInputAttachmentIndexInfoKHR *dynInputAttachmentIndex =
+            (const VkRenderingInputAttachmentIndexInfoKHR *)FindNextStruct(
+                BeginInfo.pInheritanceInfo,
+                VK_STRUCTURE_TYPE_RENDERING_INPUT_ATTACHMENT_INDEX_INFO_KHR);
+        m_BakedCmdBufferInfo[BakedCommandBuffer].state.dynamicRendering.localRead.Init(
+            dynAttachmentLocations, dynInputAttachmentIndex);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Why is this needed during loading? we generally don't track full state here. E.g. note the other members from `pInheritanceInfo` are not used in this path unlike the active replaying path above.

File: renderdoc/driver/vulkan/vk_info.h [hunk 10]
```diff
@@ -183,6 +183,47 @@ struct DescUpdateTemplate
   rdcarray<VkDescriptorUpdateTemplateEntry> updates;
 };
 
+struct DynamicRenderingLocalRead
+{
+  void Init(const VkRenderingAttachmentLocationInfoKHR *attachmentLocations,
+            const VkRenderingInputAttachmentIndexInfoKHR *inputAttachmentIndex);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_info.h]: I think given the uses of this it might make more sense to have this take a base `pNext` struct (`VkBaseInStructure`) and do the pNext search itself, since that's what each call does. Any use which doesn't fit that can always call `UpdateLocations` and `UpdateInputIndices` manually.

File: renderdoc/driver/vulkan/vk_info.cpp [hunk 11]
```diff
@@ -2985,3 +2997,99 @@ void DescUpdateTemplate::Apply(const void *pData, DescUpdateTemplateApplication
       application.writes.push_back(write);
   }
 }
+
+void DynamicRenderingLocalRead::Init(const VkRenderingAttachmentLocationInfoKHR *attachmentLocations,
+                                     const VkRenderingInputAttachmentIndexInfoKHR *inputAttachmentIndex)
+{
+  if(attachmentLocations != nullptr)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_info.cpp]: Please use `NULL` not nullptr.

File: renderdoc/driver/vulkan/vk_state.h [hunk 12]
```diff
@@ -25,6 +25,7 @@
 #pragma once
 
 #include "vk_common.h"
+#include "vk_info.h"
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_state.h]: Rather than adding a new header include could you move the struct to `vk_common.h` instead if it is to be shared across multiple places.

File: renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp [hunk 13]
```diff
@@ -3566,6 +3566,147 @@ void WrappedVulkan::vkCmdSetRayTracingPipelineStackSizeKHR(VkCommandBuffer comma
...
+        }
+      }
+      else
+      {
+        commandBuffer = VK_NULL_HANDLE;
+      }
+    }
+    else
+    {
+      m_BakedCmdBufferInfo[m_LastCmdBufferID].state.dynamicRendering.localRead.UpdateLocations(
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp]: I don't think there's a need to update this state during loading. Most dynamic state does not get tracked here.

File: renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp [hunk 14]
```diff
@@ -3566,6 +3566,147 @@ void WrappedVulkan::vkCmdSetRayTracingPipelineStackSizeKHR(VkCommandBuffer comma
...
+  }
+}
+
+template <typename SerialiserType>
+bool WrappedVulkan::Serialise_vkCmdSetRenderingInputAttachmentIndicesKHR(
+    SerialiserType &ser, VkCommandBuffer commandBuffer,
+    const VkRenderingInputAttachmentIndexInfoKHR *pInputAttachmentIndexInfo)
+{
+  SERIALISE_ELEMENT(commandBuffer);
+  SERIALISE_ELEMENT_LOCAL(inputAttachmentIndex, *pInputAttachmentIndexInfo)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp]: `inputAttachmentIndexInfo` to preserve the naming better, as this is not just one index.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp [hunk 15]
```diff
@@ -41,6 +41,8 @@ Q_DECLARE_METATYPE(CombinedSamplerData);
 
 namespace
 {
+constexpr uint32_t VK_ATTACHMENT_UNUSED = ~0U;
```
Comment (baldurk) [inline qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: I'd rather make this an `UnusedLocation` constant in `VKPipe::RenderPass` so that it's explicitly documented and the magic value is surfaced for python users as well.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp [hunk 16]
```diff
@@ -2819,6 +2841,22 @@ void VulkanPipelineStateViewer::setState()
...
+                                .arg((blend.writeMask & 0x8) == 0 ? lit("_") : lit("A"));
+
+        // With VK_KHR_dynamic_rendering_local_read, if a color attachment is mapped to
+        // VK_ATTACHMENT_UNUSED, it is implicitly disabled.  The Slot name in the "Render Pass"
+        // pane already tags the attachment with [disabled], so the write mask is simply set to
+        // ____ here.
+        if(i < rp.colorAttachmentLocations.count() &&
+           rp.colorAttachmentLocations[i] == VK_ATTACHMENT_UNUSED)
+        {
+          writemask = lit("____");
```
Comment (baldurk) [inline qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: I think we should have a different display here so there's no chance of confusion about why writes are dropped. It's not really the write mask at all but I can appreciate the logic of putting something here. Setting it to `DISABLED` or `UNUSED` would be more clear I feel.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: renderdoc/api/replay/vk_pipestate.h (hunk 2 above)
Comment (ShabbyX) [reply renderdoc/api/replay/vk_pipestate.h]: Yes, the value is unused if `isDepthInputAttachmentIndexImplicit`. I set it to something invalid to catch bugs earlier, changed to UINT_MAX instead so it's less magic. (Default to 0 could do surprising things if an bug makes RD use this value when it shouldn't).

File: renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp (hunk 13 above)
Comment (ShabbyX) [reply renderdoc/driver/vulkan/wrappers/vk_dynamic_funcs.cpp]: I have to admit I didn't quite follow what exactly the backed info is used for. The location mapping is a little special as it affects things like looking up the color attachment name in the shader and maybe other things like messing up the history/overlay's usage of "attachment 0", things like that which the typical dynamic state has no bearing on.

Also, unlike other dynamic state, Vulkan allows the app to pass this in the secondary command buffer's inheritance info, which makes me wonder if it has to have a special connection to the CB (like being baked in it? :stuck_out_tongue:)

I'll trust your judgement here, if you could please verify that this is definitely useless.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp (hunk 15 above)
Comment (ShabbyX) [reply qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: Ok, this is a Vulkan constant, I renamed it (not to conflict with Vulkan's `#define`) and put it there. Is `static constexpr` ok, or should this be a normal member for it to work in Python? I went with normal for now, but `static constexpr` would make more sense.

File: renderdoc/driver/vulkan/vk_pixelhistory.cpp (hunk 6 above)
Comment (ShabbyX) [reply renderdoc/driver/vulkan/vk_pixelhistory.cpp]: The reason I did this is because DRLR may not always be available. If the values are non-default, it means DRLR was definitely used and the extension was available.

Would you prefer I replace this with an extension check?

File: renderdoc/driver/vulkan/vk_replay.cpp (hunk 8 above)
Comment (ShabbyX) [reply renderdoc/driver/vulkan/vk_replay.cpp:1955]: They are already reset! `ret` is completely reset at the top of the function:

```
    ret = VKPipe::State();
```

and these values have default initializers. Some of the other stuff set in `else` are redundant in the current code.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 9 above)
Comment (ShabbyX) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: :shrug: (lack of holistic view of RD's codebase)

There was another comment about baked info too, let me play with that a bit and make sure removing this (and the one in the other comment) don't break anything.

File: renderdoc/driver/vulkan/vk_stringise.cpp [hunk 17]
```diff
@@ -207,6 +207,8 @@ rdcstr DoStringise(const VulkanChunk &el)
     STRINGISE_ENUM_CLASS(vkCmdSetVertexInputEXT)
     STRINGISE_ENUM_CLASS(vkCmdBeginRendering)
     STRINGISE_ENUM_CLASS(vkCmdEndRendering)
+    STRINGISE_ENUM_CLASS(vkCmdSetRenderingAttachmentLocationsKHR)
+    STRINGISE_ENUM_CLASS(vkCmdSetRenderingInputAttachmentIndicesKHR)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_stringise.cpp]: These stringise cases should be re-ordered to match the enum declaration order.

File: renderdoc/driver/vulkan/vk_pixelhistory.cpp (hunk 6 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_pixelhistory.cpp]: To be clear the current code is correct, I only mention this as a simplification to potentially make things a bit less error-prone with future refactors.

Because we are disabling the extension use and setting it back to default this will work out neatly regardless of if the extension is available or used, without needing any check - as you say if the values are non-default then the extension was used and we can restore it, but the reverse is also true that if the extension was not used then the values will be default.

As an analogy consider a new extension state `VkFrobulator` which can be set to `VK_FROBULATE` or `VK_EXTRA_FROBULATE` in an extension and we want to turn it off for the primitive ID pipe. This code:

```cpp
if(isPrimitiveIDPipe && prevState.frobulator != VK_DONT_FROTBULATE)
  state.frobulator = VK_DONT_FROTBULATE;

// ... ReplayDraw etc

if(isPrimitiveIDPipe && prevState.frobulator == VK_FROBULATE)
  state.frobulator = prevState.frobulator;
```

will be equivalent to:

```cpp
if(isPrimitiveIDPipe)
  state.frobulator = VK_DONT_FROTBULATE;

// ... ReplayDraw etc

state.frobulator = prevState.frobulator;
```

and it will work safely. If the extension was not enabled, the state was already VK_DONT_FROTBULATE so we are only doing an identity assignment multiple times but at no time do we potentially use the extension when it's not available. It is certainly redundant, but doesn't need any checks at all - we only need a check when we are _enabling_ use of an extension which might not be present.

File: renderdoc/api/replay/vk_pipestate.h [hunk 18]
```diff
@@ -859,6 +859,34 @@ samples used to render this subpass.
...
+  rdcarray<uint32_t> colorAttachmentInputIndices;
+
+  DOCUMENT("Whether or not depth input attachment index is implicit (dynamic rendering).");
+  bool isDepthInputAttachmentIndexImplicit = true;
+
+  DOCUMENT("Whether or not stencil  input attachment index is implicit (dynamic rendering).");
+  bool isStencilInputAttachmentIndexImplicit = true;
+
+  DOCUMENT("Depth input attachment index if explicit (dynamic rendering).");
+  uint32_t depthInputAttachmentIndex = UINT32_MAX;
```
Comment (baldurk) [inline renderdoc/api/replay/vk_pipestate.h:888]: By convention I usually use `~0U` for these rather than `UINT32_MAX`. Means the same thing but it matches the other code.

File: renderdoc/driver/vulkan/vk_replay.cpp (hunk 8 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_replay.cpp:1955]: Ahh I clearly forgot that that's there, my bad! The other APIs don't do that and instead update in-place since these state structures can be quite large. It looks like a potential mistake looking at the history - I might remove that in future then but for now this is totally fine.

File: renderdoc/api/replay/vk_pipestate.h [hunk 19]
```diff
@@ -859,6 +859,34 @@ samples used to render this subpass.
...
+
+  DOCUMENT("Stencil input attachment index if explicit (dynamic rendering).");
+  uint32_t stencilInputAttachmentIndex = UINT32_MAX;
+
+  DOCUMENT(R"(Alias for VK_ATTACHMENT_UNUSED, for use by the UI to know when a value in
+colorAttachmentLocations or colorAttachmentInputIndices is mapped to VK_ATTACHMENT_UNUSED.)");
+  uint32_t vkAttachmentUnused = ~0U;
```
Comment (baldurk) [inline renderdoc/api/replay/vk_pipestate.h]: _Forking slightly from the [other comment](https://github.com/baldurk/renderdoc/pull/3459#discussion_r1818962786) since github's PR view is not very good and only allows me to reply to a conversation with a review comment if the conversation still exists in the diff_

This should be `static const` rather than a normal member, which will be picked up by python. Also please drop the 'vk' prefix as it's redundant in this context. The only difference to account for is that the docstring has to be placed in the class docstring. Look at `IReplayController::`NoPreference` as an example of how this can look.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 20]
```diff
@@ -1258,6 +1258,9 @@ bool WrappedVulkan::Serialise_vkBeginCommandBuffer(SerialiserType &ser, VkComman
               BeginInfo.pInheritanceInfo->subpass;
           // framebuffer is not useful here since it may be incomplete (imageless) and it's
           // optional, so we should just treat it as never present.
+
+          m_BakedCmdBufferInfo[BakedCommandBuffer].state.dynamicRendering.localRead.Init(
+              (const VkBaseInStructure *)BeginInfo.pInheritanceInfo);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:1263]: Forking slightly from other conversations to expand a bit on what this is doing and what is needed for state tracking. You can skip this comment if you want, as it's just informational to explain what you were unsure about and nothing needs to change as the code looks right now.

---

The idea of baked command buffers is related to the command buffer lifecycle itself and versioning. Imagine a `command buffer 1234` which is created, in the normal API this is recorded, then submitted 1 to N times, then potentially reset and used again. While using this 'live' there's no ambiguity about what that `command buffer 1234` handle refers to, but when replaying a capture there can be ambiguity if the command buffer is reset an re-recorded multiple times across the scope of the capture.

To address this RenderDoc has the concept of 'baked' command buffers. Whenever you begin/end `command buffer 1234` that particular _version_ of it gets treated as if a new object was created `baked command buffer 1235` which is immutable and only used whenever as that set of recorded commands is. If there is a rerecord, that becomes `baked command buffer 1236`. You can see this in the event browser if you look as command buffers will be referred to as 'baked'. There are some more details around ensuring the right version is used at the right time but that's not needed to understand this.

Given that, the baked command buffer info is not "baked (command buffer info)" but rather "(baked command buffer) info". Each version - of which there is almost always just one anyway - gets its own info so that its state, submissions, events etc can all be tracked properly.

Regarding the state tracking, generally we don't need to track any state during the initial load because we're replaying the whole capture and we don't need to care about the state along the way for displaying in the UI or for doing partial-replays of command buffers. We _do_ end up tracking a subset of state to be able to construct the information about resource 'usage in frame' - namely descriptor set and framebuffer bindings so that we know which images are used for what at any given event. That's why the `BeginRendering` dynamic rendering functions will be doing some state tracking that you might have seen, but we don't need to track anything that doesn't affect which resources are used like dynamic depth state or in this case a dynamic permutation of attachments.

---

Hopefully that makes sense as an info dump, like I say this all looks good now.

**Review by baldurk — CHANGES_REQUESTED** [review]
This looks good now - I've left a couple of minor tweaks remaining and one explanation of the command buffer things that you were unsure of in case you want the info.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp [hunk 21]
```diff
@@ -2614,12 +2614,32 @@ void VulkanPipelineStateViewer::setState()
...
+          // example when specifying blend state for attachments or with vkCmdClearAttachments. As
+          // such, RenderDoc shows the same index in Color attachments (i.e. fbIdx == localIdx) to
+          // avoid confusion, even when VK_KHR_dynamic_rendering_local_read maps these attachments
+          // to different "locations" used by the shader.  In that case, the mapped location is
+          // shown besides the attachment index.
+          uint32_t location = a.localIdx;
+          if(a.fbIdx < rp.colorAttachmentLocations.count())
+          {
+            location = rp.colorAttachmentLocations[a.fbIdx];
+            if(location == rp.AttachmentUnused)
```
Comment (baldurk) [inline qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: One more minor nit, accessing a constant like this makes it read like it's a real member which could be confusing. I'd rather you do explicitly `VKPipe::RenderPass::AttachmentUnused` to be very clear that it's a constant.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp (hunk 21 above)
Comment (ShabbyX) [reply qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: Done
